### PR TITLE
FlightTask Introduction Replace Legacy Multicopter Position Controller

### PR DESCRIPTION
**Pull request by MaEtUgR**
It's a follow up to #9253 and #9368.

I rebased all changes to any FlightTasks and the PositionControl logic and left out all the changes to the `mc_pos_control_main.cpp` file because of nasty rebase conflicts. Then in the last commit "mc_pos_control: replace legacy functionality" I switched the mc_pos_control file to the new refactored version that does not contain any unwanted, replaced legacy code anymore.
Note: The original changes are mostly by @Stifael as correctly represented in the metadata.

We know what's missing right now is:

- [x] Porting changes that happened to the position controller in the meantime to FlightTasks
      See any of these sources:
      - [History in June](https://github.com/PX4/Firmware/commits/master/src/modules/mc_pos_control/mc_pos_control_main.cpp)
      - The file `src/modules/mc_pos_control/mc_pos_control_main.cpp`
in this diff: https://github.com/PX4/Firmware/compare/Legacy-Auto-Manual-replacement...master
      - The only prs that did changes: https://github.com/PX4/Firmware/pull/9613 and https://github.com/PX4/Firmware/pull/9630

closes #9368

### CODE REVIEW

**Comment by dagar** [comment]
Given https://github.com/PX4/Firmware/pull/9270#issuecomment-399362836 is it worth pushing the core FlightTasks interface improvements to master in a separate PR? I'm wondering if that would help unblock obstacle avoidance migration.

**Comment by Stifael** [comment]
@dagar, why would obstacle avoidance be blocked by Flighttask?

**Comment by Stifael** [comment]
@priseborough @mhkabir 
I ported the mc_pos_control_main changes from #9613 to flighttask (commits: 
https://github.com/PX4/Firmware/pull/9731/commits/8405d90f468bf972c3e6908b270060d16044d28c, 
https://github.com/PX4/Firmware/pull/9731/commits/a10a9a3366381a9a0e6ba0a54f10bc6d3dd2ad50,
https://github.com/PX4/Firmware/pull/9731/commits/34c881caab7604db9427628c9541370cb5ed7266).

A couple questions about the desired distance to ground behavior:
* if MPC_ALT_MODE is 0:
   * Is the range sensor used for landing and takeoff only as explained [here](https://github.com/PX4/Firmware/pull/9613/files#diff-573b034156257a95a309dedc0606f637R1078)? If yes, why is then there an altitude limit based on distance to ground during normal altitude mode [here](https://github.com/PX4/Firmware/pull/9613/files#diff-fc77c3ef569029d45764664c75d4b0c1R856). 
   * what are the scenarios when ekf2 sets the `dist_to_bottom_valid` flag to false? if outside the hagl_min/max?

File: src/modules/mc_pos_control/mc_pos_control_main.cpp [hunk 1]
```diff
@@ -770,2524 +369,510 @@ MulticopterPositionControl::task_main_trampoline(int argc, char *argv[])
...
+			if (!_flight_tasks.update()) {
+				// FAILSAFE
+				// Task was not able to update correctly. Do Failsafe.
+				setpoint.x = setpoint.y = setpoint.z = NAN;
+				setpoint.vx = setpoint.vy = setpoint.vz = NAN;
+				setpoint.thrust[0] = setpoint.thrust[1] = setpoint.thrust[2] = NAN;
 
-	/* check if stick direction and current velocity are within 60angle */
-	const bool is_aligned = (stick_xy_norm * stick_xy_prev_norm) > 0.5f;
+				if (PX4_ISFINITE(_states.velocity(2))) {
```
Comment (Stifael) [inline src/modules/mc_pos_control/mc_pos_control_main.cpp]: move this stuff outside. this requires to check for landdetector and  if in air, then slowly descend, if landed, then just go idle

**Comment by mhkabir** [comment]
@Stifael about your questions :

File: src/modules/mc_pos_control/mc_pos_control_main.cpp [hunk 2]
```diff
@@ -770,2524 +369,510 @@ MulticopterPositionControl::task_main_trampoline(int argc, char *argv[])
...
-	/* compute velocity target close to waypoint */
-	float vel_close;
+	} else if (PX4_ISFINITE(_local_pos.vz)) {
 
-	if (use_linear_approach) {
+		if (PX4_ISFINITE(vel_sp_z) && fabsf(vel_sp_z) > FLT_EPSILON && PX4_ISFINITE(_local_pos.z_deriv)) {
+			// A change in velocity is demanded. Set velocity to the derivative of position
+			// because it has less bias but blend it in across the landing speed range
+			float weighting = fminf(fabsf(vel_sp_z) / _land_speed.get(), 1.0f);
+			_states.velocity(2) = _local_pos.z_deriv * weighting + _local_pos.vz * (1.0f - weighting);
```
Comment (RomanBapst) [inline src/modules/mc_pos_control/mc_pos_control_main.cpp]: @Stifael @MaEtUgR Aren't you overwriting this again a few lines below?

File: src/modules/mc_pos_control/mc_pos_control_main.cpp [hunk 3]
```diff
@@ -770,2524 +369,510 @@ MulticopterPositionControl::task_main_trampoline(int argc, char *argv[])
...
+		if (fabsf(setpoint.vz) * _dt > delta_p) {
+			setpoint.z = -_vehicle_land_detected.alt_max +  _home_pos.z;
+			setpoint.vz = 0.0f;
+		}
 	}
 }
 
 void
-MulticopterPositionControl::reset_alt_sp()
+MulticopterPositionControl::check_vehicle_states(const float &vel_sp_z)
```
Comment (RomanBapst) [inline src/modules/mc_pos_control/mc_pos_control_main.cpp]: @MaEtUgR @Stifael Maybe call this function set_vehicle_states()?

File: src/modules/mc_pos_control/mc_pos_control_main.cpp [hunk 4]
```diff
@@ -770,2524 +369,510 @@ MulticopterPositionControl::task_main_trampoline(int argc, char *argv[])
...
-float
-MulticopterPositionControl::get_cruising_speed_xy()
-{
-	/*
-	 * in mission the user can choose cruising speed different to default
-	 */
-	return ((PX4_ISFINITE(_pos_sp_triplet.current.cruising_speed) && !(_pos_sp_triplet.current.cruising_speed < 0.0f)) ?
-		_pos_sp_triplet.current.cruising_speed : _vel_cruise_xy.get());
+	if (PX4_ISFINITE(_local_pos.yaw)) {
+		_states.yaw = _local_pos.yaw;
```
Comment (RomanBapst) [inline src/modules/mc_pos_control/mc_pos_control_main.cpp]: @MaEtUgR @Stifael We need to check is not isfinite implies that it is NaN. Isn't ther also stuff like INFINITY?

File: src/modules/mc_pos_control/mc_pos_control_main.cpp [hunk 5]
```diff
@@ -770,2524 +369,510 @@ MulticopterPositionControl::task_main_trampoline(int argc, char *argv[])
...
+				setpoint.yawspeed = NAN;
+				setpoint.yaw = _states.yaw;
+				constraints.landing_gear = vehicle_constraints_s::GEAR_KEEP;
+			}
 
-	} else if (do_deceleration) {
-		/* we do manual deceleration */
-		intention = deceleration;
+			// limit altitude only if local position is valid
+			if (PX4_ISFINITE(_states.position(2))) {limit_altitude(setpoint);}
```
Comment (RomanBapst) [inline src/modules/mc_pos_control/mc_pos_control_main.cpp]: @MaEtUgR @Stifael I was confused a while

File: src/modules/mc_pos_control/mc_pos_control_main.cpp [hunk 6]
```diff
@@ -770,2524 +369,510 @@ MulticopterPositionControl::task_main_trampoline(int argc, char *argv[])
...
+			q_sp.copyTo(_att_sp.q_d);
+			_att_sp.q_d_valid = true;
+			_att_sp.thrust = 0.0f;
+		}
+	}
 
-		if (_jerk_hor_max.get() > _jerk_hor_min.get()) {
-			_manual_jerk_limit_xy = (_jerk_hor_max.get() - _jerk_hor_min.get()) / _velocity_hor_manual.get() *
-						sqrtf(_vel(0) * _vel(0) + _vel(1) * _vel(1)) + _jerk_hor_min.get();
+	mavlink_log_info(&_mavlink_log_pub, "[mpc] stopped");
```
Comment (RomanBapst) [inline src/modules/mc_pos_control/mc_pos_control_main.cpp]: @MaEtUgR @Stifael I think we can get rid of this.

File: src/modules/mc_pos_control/mc_pos_control_main.cpp (hunk 4 above)
Comment (Stifael) [reply src/modules/mc_pos_control/mc_pos_control_main.cpp]: INFINITY and NAN should both return 0 is checked with PX4_ISFINITE.

File: src/modules/mc_pos_control/mc_pos_control_main.cpp (hunk 2 above)
Comment (Stifael) [reply src/modules/mc_pos_control/mc_pos_control_main.cpp]: fixed

**Comment by MaEtUgR** [comment]
I rebased again in the hope CI will be happier.

**Comment by Stifael** [comment]
Here is again the summary of the PR: 
https://github.com/PX4/Firmware/pull/9368#issuecomment-386545525

Here is again the summary of the different files:
https://github.com/PX4/Firmware/pull/9368#issuecomment-393058164

What needs to be tested: 
https://github.com/PX4/Firmware/pull/9368#issuecomment-392529945

**Comment by Stifael** [comment]
@dagar sounds good.
But once this document is finished, one really need to compare it to the legacy code. This PR does not fix what is already broken except of some smaller stuff that were obvious to fix. For instance, some parts of follow-me is not executed already.

**Comment by dagar** [comment]
That's fine, we really just need to have a thorough understanding of the state before and after. Functionality doesn't necessarily need to be replicated 100%. For example if follow me doesn't actually work in master we might be better off leaving it temporarily disabled in FlightTasks until someone wants to get it working properly.

File: src/modules/mc_pos_control/mc_pos_control_main.cpp [hunk 7]
```diff
@@ -319,97 +165,70 @@ class MulticopterPositionControl : public control::SuperBlock, public ModulePara
...
+	 * Thrust is adjusted to support the land-detector during detection.
+	 * @param thrust_setpoint gets adjusted based on land-detector state
 	 */
-	void updateConstraints(Controller::Constraints &constrains);
+	void limit_thrust_during_landing(matrix::Vector3f &thrust_sepoint);
 
-	void publish_attitude();
+	/**
+	 * Start flightasks based on navigation state.
+	 * This methods activates a taks basedn on the navigation state.
```
Comment (mrivi) [inline src/modules/mc_pos_control/mc_pos_control_main.cpp]: `taks basedn` -> `task based`

**Comment by Stifael** [comment]
@PX4/testflights 
can you do flight-tests based on the document [here](https://docs.google.com/document/d/1q1wkqufilFYRzskx9OJcmOHpEQOAgs8fEKEHX2V4ooo/edit)?

**Comment by santiago3dr** [comment]
some flights from a pixracer (V4)

Stabilized, Altitude and Position mode:
https://logs.px4.io/plot_app?log=952d9697-5a02-4469-aa55-fa4a0eb5315a
https://logs.px4.io/plot_app?log=c6fa2f0b-1d89-4cd7-b640-235857077f6c

Tighter yaw response; great flight
same flight from master for comparison:
https://logs.px4.io/plot_app?log=8dc45147-f0c2-4e51-a53e-0db154978d17

Auto take-off and RTL as part of mission with second waypoint including delay:
https://logs.px4.io/plot_app?log=0716dd29-5e27-4319-864a-129bf57dda54
https://logs.px4.io/plot_app?log=4f7996e5-923a-4b79-8e5c-20eb44885ced

Slight delay upon take-off and a disarm taking a couple seconds longer upon landing

same mission running master with immediate take-off upon engaging Auto for comparison:
https://logs.px4.io/plot_app?log=daae9d74-f080-422b-8dbc-b99e6cf5d808

Auto take-off and last waypoint set to LAND instead of RTL:
https://logs.px4.io/plot_app?log=a5d0cca4-e599-461c-83fb-90864082f29c
https://logs.px4.io/plot_app?log=f67bbf5e-5700-471c-b93c-d46da4fc8f6a
Slight delay upon take-off and a disarm taking a couple seconds longer upon landing

Overall good flights

**Comment by Stifael** [comment]
for follow-up PRs (things that are not important for this PR but will improve flighttask)
TODO:

- [ ] replace setpoint memebers with setters and getters to make it clear that there must be a valid setpoint in the xy-direction and a valid in z-direction.

**Comment by ChristophTobler** [comment]
@MaEtUgR @Stifael 
As discussed, I fixed the position control failsafe (last one to trigger) and added a failsafe if all position, velocity and thrust pairs are invalid.

**Comment by Avysuarez** [comment]
Some flights with pixhawk 4 (V5)
stabilized, altitude and position mode: 
https://review.px4.io/plot_app?log=1847a1d8-1f0c-4186-ba06-e37ae5f0ab48
https://review.px4.io/plot_app?log=70d16986-e748-45af-a9cc-3da998d7f323
Good flights no issues.

Full auto:
https://review.px4.io/plot_app?log=14618afe-88b0-4b8a-b891-a77313407fbf
https://review.px4.io/plot_app?log=0d3bd44e-1cc7-454b-a8ac-2f725b16d548
Good flight, but has a delay in take-off and when it lands it takes like a 15 seconds to disarm.

Position mode, auto mode and RTL.
https://review.px4.io/plot_app?log=075e0e71-5456-4f24-bbcf-266bede54c0b
https://review.px4.io/plot_app?log=c7aadcd7-1243-4171-b751-fc354dc1c7a9
Good flight, but has a delay in the throttle for take-off and when it lands it takes like a 15 seconds to disarm.

**Comment by dannyfpv** [comment]
test flight pixhack v5 
https://logs.px4.io/plot_app?log=19efaa4d-d651-4a5e-9f0b-30b7a478bdde

pixhawk pro v4
https://logs.px4.io/plot_app?log=dc03db78-47e6-498c-ba4f-f59da8451020
delay up on take-off and a disarm

**Comment by Stifael** [comment]
commit e3f92e9 removed the delay for takeoff. However, the delay can still be enabled through MPC_IDLE_TKO, which can be useful for larger vehicles

**Comment by MaEtUgR** [comment]
@santiago3dr Thanks for the detailed report!!

> Tighter yaw response; great flight

Nice to see that the efforts to make yaw control more precise resulted in a well noticable improvement.

> Good flight, but has a delay in take-off and when it lands it takes like a 15 seconds to disarm.

Delay in take off is intentionally added by @Stifael to make sure aricrafts with big propellers have enough time to spin up. Just now it was made optional with the parameter `MPC_IDLE_TKO` because we agree that it's not useful for everyone. Landing detection taking longer is an unknown side effect we need to check. Did the slow landing happen with multiple vehicles?

@Stifael 
> replace setpoint memebers with setters and getters to make it clear that there must be a valid setpoint in the xy-direction and a valid setpoint in z-direction.

I'm against reverting the architecture to setters and getters for the setpoints again. Initially it was the way I started out but we already had numerous discussion why that was not a scalable diecision. I would leave the interface, document it well and make sure the failsafes work correctly (which is already the case according to @ChristophTobler ).

File: src/lib/FlightTasks/tasks/FlightTaskAuto.cpp [hunk 8]
```diff
@@ -0,0 +1,254 @@
...
+		map_projection_init(&_reference_position,
+				    _sub_vehicle_local_position->get().ref_lat,
+				    _sub_vehicle_local_position->get().ref_lon);
+		_reference_altitude = _sub_vehicle_local_position->get().ref_alt;
+		_time_stamp_reference = _sub_vehicle_local_position->get().ref_timestamp;
+	}
+
+	if (PX4_ISFINITE(_reference_altitude)
+	    && PX4_ISFINITE(_sub_vehicle_local_position->get().ref_lat)
+	    && PX4_ISFINITE(_sub_vehicle_local_position->get().ref_lat)) {
```
Comment (MaEtUgR) [inline src/lib/FlightTasks/tasks/FlightTaskAuto.cpp]: copy pasta: `ref_lon`

**Comment by Stifael** [comment]
@santiago3dr @Avysuarez havent figure out exactly why the thrust takes for ever to drop below the threshold during landing. Was the ground during landing flat or inclined?

**Comment by Avysuarez** [comment]
@Stifael 
In my case flat and in all the flights in auto landing it took more time to disarm.

**Comment by Stifael** [comment]
@Avysuarez I  know now why the landing behaves differently on master, and it turns out that it is a bug on master, which accelerates the landing. 
The intended landing procedure consists of 3 states:
- ground contact: turn off thrust horizontally
- maybe landed: turn off thrust in z
- landed: vehicle is not moving despite zero thrust

In the legacy controller, however, ground contact already turns off thrust in all three direction: https://github.com/PX4/Firmware/blob/master/src/modules/mc_pos_control/mc_pos_control_main.cpp#L3631

```
thrust_sp_body(2) = thrust_sp(2) > 0.0f ? thrust_sp(2) : 0.0f;
```

This line means that if thrust setpoint in z-direction is larger than 0, then keep the original thrust septoint, otherwise set it to 0. However, thrust setpoint in z-direction is ALWAYS smaller than 0 for any altitude controlled mode because of NED frame and therefore `thrust_sp(2)` is set to 0 already during ground contact. This is a bug on master.

The landdetection procedure is implemented conservative on purpose but should still detect landing within a reasonable amount of time. However, it depends a lot on the parameters.

I saw that you used some of the default parameters:
MPC_THR_HOVER = 0.5
MPC_Z_VEL_P = 0.2
MPC_Z_VEL_I = 0.02

These values influence the detection a lot. If hover thrust is set too low, then `MPC_Z_VEL_I` will correct for it by making the integral of the error more negative with a rate set by `MPC_Z_VEL_I`. This means that if `MPC_Z_VEL_I` is small, then it will take some time to correct for the wrong hover thrust. The actual problem, however, arises during landing, where the large accumulated, negative integral needs to drop (become less negative). With a badly set `MPC_THR_HOVER` and a low `MPC_Z_VEL_I`, the integral drop will take for ever and therefore landing will be delayed. 

@bkueng just recently updated the tuning page (https://docs.px4.io/en/advanced_config/pid_tuning_guide_multicopter.html). I think it would be worth to change the default values based on the procedure described in the tuning page.

**Comment by Avysuarez** [comment]
@Stifael Thank you!

**Comment by Stifael** [comment]
@PX4/testflights do you mind to test this branch once more, just to get some more flight hours, which should bring us closer to a version that we can finally merge.

**Comment by Avysuarez** [comment]
Some flights with pixhawk 2.1 (V3)

stabilized, altitud and position mode.
https://review.px4.io/plot_app?log=655a41b7-23c5-488d-b734-243ecb8b2be5
https://review.px4.io/plot_app?log=160409b8-db58-4280-827a-b9f0db25e44b

Full auto flights.
https://review.px4.io/plot_app?log=25ef8c45-9c43-4154-bff7-3d695a1a0868
https://review.px4.io/plot_app?log=86fd0a50-8e59-4ac3-9282-5fa97720626b

position, auto and RTH.
https://review.px4.io/plot_app?log=4d9ecbc1-7d76-40ca-9d98-0e7e9f5e7dea
https://review.px4.io/plot_app?log=82bd68ed-f9c4-48a6-9aa7-39e171d9958d
GOOD FLIGHTS, NO ISSUES.

Some flights with pixhawk 4 (V5)

stabilized, altitude and position mode.
https://review.px4.io/plot_app?log=13ca31c7-b720-4c61-8941-b9ebb6ee7de5
https://review.px4.io/plot_app?log=36f2c5a1-0f5d-4e02-8ab8-3b2cafb62d31

Full auto flights.
https://review.px4.io/plot_app?log=51af2ea3-663d-4d81-8fb0-7b673eb9953d
https://review.px4.io/plot_app?log=1fb8062d-60a3-4556-b867-93286ec0bded

position, auto and RTH.
https://review.px4.io/plot_app?log=79fe8fae-131a-410f-8516-d2f95928dd93
https://review.px4.io/plot_app?log=983b0ff7-3e41-4fa7-b9c6-e922552385c1
GOOD FLIGHTS NO ISSUES.

**Comment by santiago3dr** [comment]
Flew on a Pixracer (V4) with new commits
manual flights went well, with the same improved yaw response:
https://logs.px4.io/plot_app?log=fd543ccb-4dea-4f6f-b9b9-8c45d9b68868
https://logs.px4.io/plot_app?log=ad00b7d7-48b9-4de7-9ba9-a40cf524097b

Played around with MPC_VEL_P and MPC_VEL_I to get a better landing
MPC_VEL_P = 1.0 (default)
MPC_VEL_I = 0.02 (default)
https://logs.px4.io/plot_app?log=59dde206-28a1-476a-84da-13904cc92c87
https://logs.px4.io/plot_app?log=151b9d10-8045-446f-a9f6-c33a0076d41f

MPC_VEL_P = 1.25
MPC_VEL_I = 0.025
https://logs.px4.io/plot_app?log=2372fc9d-5449-499e-8758-1c1ec9376403
https://logs.px4.io/plot_app?log=d9d2e9b6-332e-43f0-b5f1-1e4d4fbe247a

MPC_VEL_P = 1.5
MPC_VEL_I = 0.03
https://logs.px4.io/plot_app?log=45e3b22f-024c-4ec6-8ffd-03112731c52b
https://logs.px4.io/plot_app?log=b8fa2832-10e8-4b3e-a3c3-06b5cbf626e0

MPC_VEL_P = 1.25
MPC_VEL_I = 0.03
https://logs.px4.io/plot_app?log=2b1ada59-46a5-4402-826f-be3ef03bdf78
https://logs.px4.io/plot_app?log=098fdec7-dce4-48cc-9695-a9225112c087

MPC_VEL_P = 1.25
MPC_VEL_I = 0.035
https://logs.px4.io/plot_app?log=5e9184ca-f9a1-4377-8b2d-3bd673c9c56e
https://logs.px4.io/plot_app?log=a0d1d358-0cbf-4867-8a7d-400a1f1986f7

I felt the best values were MPC_VEL_P = 1.25 and MPC_VEL_I = 0.03
If there are additional values someone could suggest, i can test them out

**Comment by LorenzMeier** [comment]
@dagar AV-X target failed: `Makefile:154: recipe for target 'av-x-v1_default' failed`

**Comment by dannyfpv** [comment]
Flew on a Pixhack (V5) with new commits
Full auto flight.
https://review.px4.io/plot_app?log=df816e7c-0bdc-41a1-8561-77e229eb25df

stabilized, altitud and position mode.
https://review.px4.io/plot_app?log=c79567c9-2985-4c99-8379-522d5193ed50

Flew on a Pixmini (V3) with new commits
Full auto flight.
https://review.px4.io/plot_app?log=0cc6fbab-58d7-494c-98ff-d6d3d30517d4
stabilized, altitud and position mode.
https://review.px4.io/plot_app?log=c1918a39-18b9-4a64-a397-6ddcf9b0dc93

Flew on a Pixhawk pro (V4) with new commits
 auto flight, position mode.
https://review.px4.io/plot_app?log=22eed7b3-6d19-47dd-8e3d-2fbde1545dd5
stabilized, rc fail safe
https://review.px4.io/plot_app?log=db2b1af2-9c19-48f6-b559-ec21bcb8afd2

**Pull request merged**